File: app/src/main/java/ch/bailu/aat/map/MapColor.java
package ch.bailu.aat.map;


import android.graphics.Color;

import ch.bailu.aat.gpx.InfoID;
import ch.bailu.aat.util.ui.AppTheme;
import ch.bailu.aat.util.ui.UiTheme;

public final class MapColor {

    private static final UiTheme theme = AppTheme.bar;


    public static final int ALPHA_HIGH = 50;
    public static final int ALPHA_LOW = 200;



    public final static int LIGHT = setAlpha(Color.WHITE, ALPHA_LOW);
    public final static int MEDIUM = setAlpha(Color.BLACK, ALPHA_HIGH);
    public final static int DARK = setAlpha(Color.BLACK, ALPHA_LOW);




    public final static int NODE_NEUTRAL = setAlpha(Color.LTGRAY, ALPHA_LOW);
    public final static int NODE_SELECTED = LIGHT;


    public final static int GRID = Color.GRAY;
    public final static int EDGE = Color.BLACK; //AppTheme.getAltBackgroundColor();
    public final static int TEXT = Color.BLACK;
    //public static final int LEGEND_TEXT = Color.BLACK;


    static public int setAlpha(int color, int alpha) {
        return Color.argb(alpha, Color.red(color), Color.green(color), Color.blue(color));
    }


    public static int setValue(int color, float value) {
        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        hsv[2] = value;
        return Color.HSVToColor(hsv);
    }


    public static int setSaturation(int color, float saturation) {
        float[] hsv = new float[3];
        Color.colorToHSV(color, hsv);
        hsv[1] = saturation;
        return Color.HSVToColor(hsv);
    }



    public static int getColorFromIID(int iid) {
        final int OVERLAY_COUNT = AppTheme.OVERLAY_COLOR.length;

        if (iid== InfoID.TRACKER)
            return theme.getHighlightColor();

        if (iid >= InfoID.OVERLAY && iid < InfoID.OVERLAY + OVERLAY_COUNT) {
            int slot =  iid - InfoID.OVERLAY;

            return AppTheme.OVERLAY_COLOR[slot];
        }

        if (iid == InfoID.EDITOR_DRAFT)
            return AppTheme.COLOR_GREEN;

        if (iid == InfoID.EDITOR_OVERLAY)
            return Color.MAGENTA; //AppTheme.getHighlightColor3();


        if (iid == InfoID.FILEVIEW)
            return AppTheme.COLOR_BLUE;



        return theme.getHighlightColor();
    }


    public static int toDarkTransparent(int color) {

        color = setValue(color, 0.30f);
        color = setAlpha(color, ALPHA_LOW);
        return color;
    }

    public static int toLightTransparent(int color) {

        color = setSaturation(color, 0.25f);
        color = setAlpha(color, ALPHA_LOW);
        return color;
    }

}

File: aat-lib/src/main/java/ch/bailu/aat_lib/map/MapColor.kt
package ch.bailu.aat_lib.map

import ch.bailu.aat_lib.app.AppColor
import ch.bailu.aat_lib.gpx.information.InfoID
import ch.bailu.aat_lib.lib.color.ARGB
import ch.bailu.aat_lib.lib.color.HSV

object MapColor {
    private const val ALPHA_HIGH = 50
    const val ALPHA_LOW = 200
    val LIGHT = setAlpha(ARGB.WHITE, ALPHA_LOW)
    val MEDIUM = setAlpha(ARGB.BLACK, ALPHA_HIGH)
    val DARK = setAlpha(ARGB.BLACK, ALPHA_LOW)

    @JvmField
    val NODE_NEUTRAL = setAlpha(ARGB.LTGRAY, ALPHA_LOW)
    @JvmField
    val NODE_SELECTED = LIGHT

    const val GRID = ARGB.GRAY
    const val EDGE = ARGB.BLACK
    const val TEXT = ARGB.BLACK

    fun setAlpha(color: Int, alpha: Int): Int {
        return ARGB(alpha, ARGB.red(color), ARGB.green(color), ARGB.blue(color)).toInt()
    }

    fun setValue(color: Int, value: Float): Int {
        val hsv = HSV(ARGB(color))
        hsv.setValue(value)
        return hsv.toInt()
    }

    private fun setSaturation(color: Int, saturation: Float): Int {
        val hsv = HSV(ARGB(color))
        hsv.setSaturation(saturation)
        return hsv.toInt()
    }

    fun getColorFromIID(iid: Int): Int {
        return overlayColorMap[iid]?: AppColor.HL_ORANGE
    }

    private val overlayColorMap = HashMap<Int, Int>().apply {
        put(InfoID.TRACKER, AppColor.HL_ORANGE)
        for (i in 0 until AppColor.OVERLAY_COLOR.size) {
            put(InfoID.OVERLAY + i, AppColor.OVERLAY_COLOR[i])
        }
        put(InfoID.EDITOR_DRAFT, AppColor.HL_GREEN)
        put(InfoID.EDITOR_OVERLAY, ARGB.MAGENTA)
        put(InfoID.FILE_VIEW, AppColor.HL_ORANGE)
        put(InfoID.NOMINATIM_REVERSE, AppColor.HL_OVERLAY_COLOR[0])
        put(InfoID.NOMINATIM, AppColor.HL_OVERLAY_COLOR[1])
        put(InfoID.OVERPASS, AppColor.HL_OVERLAY_COLOR[2])
        put(InfoID.POI, AppColor.HL_OVERLAY_COLOR[2])
        put(InfoID.BROUTER, AppColor.HL_OVERLAY_COLOR[3])
    }

    fun toDarkTransparent(colorIn: Int): Int {
        var colorOut = colorIn
        colorOut = setValue(colorOut, 0.30f)
        colorOut = setAlpha(colorOut, ALPHA_LOW)
        return colorOut
    }

    fun toLightTransparent(colorIn: Int): Int {
        var colorOut = colorIn
        colorOut = setSaturation(colorOut, 0.25f)
        colorOut = setAlpha(colorOut, ALPHA_LOW)
        return colorOut
    }
}
